include/ensmallen_bits/callbacks/early_stop_at_min_loss.hpp

```diff
@@ -45,11 +71,17 @@ class EarlyStopAtMinLoss
    */
   template<typename OptimizerType, typename FunctionType, typename MatType>
   bool EndEpoch(OptimizerType& /* optimizer */,
-                FunctionType& /* function */,
+                FunctionType& function,
                 const MatType& /* coordinates */,
                 const size_t /* epoch */,
-                const double objective)
+                double objective)
   {
+    if ((!predictors.empty()) and (!responses.empty()))
+    {
+      objective = network.Evaluate(predictors, responses);
```

Comment:
About testing, what we could do is to write a custom/minimal version of the existing`LogisticRegressionFunction` class and add another `Evaluate()` function that internally uses `ComputeAccuracy()` and we interpret that as loss. Does this make sense, I can elaborate on the idea.

Reply:
@zoq if I understood well, we have to write a small version of this class: https://github.com/mlpack/ensmallen/blob/82817a6e3b220cd8236121ef2ac2cb8f5e5bd32e/include/ensmallen_bits/problems/logistic_regression_function.hpp#L26
with `ComputeAccuracy()`. But I think we need to have a reference to `AnnType` in order to get to evaluate. Right?
In the actual implementation, I call `Evaluate` from `FFN`. However, in the tests, I do not think we can test ensmallen and having a dependency on mlpack. I am not sure.
A better solution is to re-open this pull request on mlpack side instead of ensmallen, this will be easier to test, but will be more confusing to users.

Reply:
Right, but it's not really necessary only for the FFN class, as long as the classifer/model implements `Evaluate(predictors, responses);` you can use the callback, correct me if I'm wrong. So what I had envision is either create a another problem based on `LogisticRegressionFunction` which implements `Evaluate(predictors, responses);` or add `Evaluate(predictors, responses);` to the exsisting one, but not sure that is possible since there already a function with the same arguments, that does something different. Actually, I don't mind to implement a fully dummy function to test the callback as well. Let me know what you think.

Reply:
@zoq I agree, what every classifier/model used that implements `Evaluate(predictors, responses);` should be able to use this callback.
It is the implementation details of `Evaluate` that I can not find a solution for it. Since adding a new `Evaluate` in the `add_evaluate.hpp` interface is not recommended, we might have to implement a new one locally (fully dummy function is a great idea). Beacause it is not the `Evaluate` function that we are testing here since whatever model is going to use this callback it will have it's own `Evaluate` that has already been tested. The main objective is to verify that the EarlyStop functionality is tested and works perfectly.

Reply:
This is what I had in mind:

```c++
class LogisticRegressionValidationFunction
{
 public:
  LogisticRegressionValidationFunction(LogisticRegression<>& lrIn, arma::mat& coordinatesIn) : lr(lrIn), coordinates(coordinatesIn)
  {
  }

  void Evaluate(arma::mat& predictors, arma::mat& responses)
  {
    lr.ComputeAccuracy(predictors, responses, coordinates);
  }

  LogisticRegression<> lr;
  arma::mat coordinates;
};

arma::mat data, testData, shuffledData;
arma::Row<size_t> responses, testResponses, shuffledResponses;

LogisticRegressionTestData(data, testData, shuffledData,
  responses, testResponses, shuffledResponses);

SMORMS3 smorms3;
LogisticRegression<> lr(shuffledData, shuffledResponses, 0.5);
arma::mat coordinates = lr.GetInitialPoint();
LogisticRegressionValidationFunction lrValidation(lr, coordinates);


EarlyStopAtMinLoss cb(lrValidation, testData, testResponses);

smorms3.Optimize(lr, coordinates, cb);
```

Reply:
Thanks, I understand, I will put everything together in new commits